File: companion/ios/CanRosettaCompanion/Remote/EdgeControlClient.swift
import Foundation

/// Discovery mode requested from the AutoPi.
///
/// - `fast`: catalog scan (known OBD PIDs / UDS DIDs).
/// - `slow`: brute-force sweep. Slower but finds more.
enum EdgeMode: String, CaseIterable, Identifiable, Sendable {
    case fast
    case slow
    var id: String { rawValue }
    var label: String { rawValue.capitalized }
}

// MARK: - Wire models (match docs/control-protocol.md exactly)
//
// All decoding uses `keyDecodingStrategy = .convertFromSnakeCase` and all
// encoding uses `.convertToSnakeCase`, so idiomatic camelCase Swift names map to
// the protocol's snake_case JSON (`sw_version` <-> `swVersion`, `t_utc` <->
// `tUtc`, `edge_utc_offset_est_s` <-> `edgeUtcOffsetEstS`, etc.).

struct HealthResponse: Decodable, Sendable {
    let ok: Bool
    let swVersion: String?
}

struct TimeResponse: Decodable, Sendable {
    let tUtc: Double
}

struct EdgeDevice: Decodable, Sendable {
    let id: String?
    let swVersion: String?
}

struct EdgeStats: Decodable, Sendable {
    let elapsedS: Double?
    let frames: Int?
    let obdSamples: Int?
}

struct DiscoverySummary: Decodable, Sendable {
    let obdPids: Int?
    let udsDids: Int?
    let plainCanIds: Int?
}

struct StatusResponse: Decodable, Sendable {
    let state: String
    let sessionId: String?
    let outputDir: String?
    let device: EdgeDevice?
    let mode: String?
    let stats: EdgeStats?
    let discoverySummary: DiscoverySummary?
    let error: String?
}

struct SessionResponse: Decodable, Sendable {
    let sessionId: String
    let outputDir: String?
    let device: EdgeDevice?
}

/// Response for `discover` / `log/start` / `log/stop` / `run` — a state change,
/// optionally with a frame count on stop.
struct CommandResponse: Decodable, Sendable {
    let state: String?
    let frames: Int?
}

/// A single line off the `GET /api/ws` event stream. All fields are optional
/// because the shape depends on `event`; we decode a superset and read what the
/// event kind supplies.
struct EdgeEvent: Decodable, Sendable {
    let event: String
    let state: String?
    let phase: String?
    let supportedPids: Int?
    let summary: DiscoverySummary?
    let frames: Int?
    let obdSamples: Int?
    let elapsedS: Double?
    let message: String?
    let ts: Double?
}

// Request bodies.

private struct SessionRequest: Encodable {
    struct Vehicle: Encodable {
        let make: String?
        let model: String?
        let year: Int?
    }
    let sessionId: String?
    let vehicle: Vehicle?
    let edgeUtcOffsetEstS: Double?
    let clockSource: String?
}

private struct DiscoverRequest: Encodable {
    let mode: String
}

private struct RunRequest: Encodable {
    let mode: String
    let durationS: Double?
}

// MARK: - Errors

/// A user-presentable error from the control client.
struct EdgeError: LocalizedError, Sendable {
    let message: String
    let statusCode: Int?
    init(_ message: String, statusCode: Int? = nil) {
        self.message = message
        self.statusCode = statusCode
    }
    var errorDescription: String? { message }
}

// MARK: - Client

/// Stateless async client for the AutoPi control protocol
/// (`docs/control-protocol.md`). Value type holding only host + token, so it is
/// trivially `Sendable` and can be recreated cheaply whenever settings change.
struct EdgeControlClient: Sendable {

    let host: String
    let token: String

    private static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }()

    private static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.keyEncodingStrategy = .convertToSnakeCase
        return e
    }()

    private var trimmedHost: String {
        var h = host.trimmingCharacters(in: .whitespacesAndNewlines)
        while h.hasSuffix("/") { h.removeLast() }
        return h
    }

    // MARK: HTTP endpoints

    func health() async throws -> HealthResponse {
        try await perform(makeRequest("/api/health", method: "GET"))
    }

    func time() async throws -> TimeResponse {
        try await perform(makeRequest("/api/time", method: "GET"))
    }

    func status() async throws -> StatusResponse {
        try await perform(makeRequest("/api/status", method: "GET"))
    }

    func createSession(
        sessionId: String?,
        edgeUtcOffsetEstS: Double?,
        clockSource: String?
    ) async throws -> SessionResponse {
        let body = SessionRequest(
            sessionId: sessionId,
            vehicle: nil,
            edgeUtcOffsetEstS: edgeUtcOffsetEstS,
            clockSource: clockSource
        )
        return try await perform(makeRequest("/api/session", method: "POST",
                                             jsonBody: try Self.encoder.encode(body)))
    }

    func discover(mode: EdgeMode) async throws -> CommandResponse {
        let body = DiscoverRequest(mode: mode.rawValue)
        return try await perform(makeRequest("/api/discover", method: "POST",
                                             jsonBody: try Self.encoder.encode(body)))
    }

    func startLog() async throws -> CommandResponse {
        try await perform(makeRequest("/api/log/start", method: "POST"))
    }

    func stopLog() async throws -> CommandResponse {
        try await perform(makeRequest("/api/log/stop", method: "POST"))
    }

    func run(mode: EdgeMode, durationS: Double? = nil) async throws -> CommandResponse {
        let body = RunRequest(mode: mode.rawValue, durationS: durationS)
        return try await perform(makeRequest("/api/run", method: "POST",
                                             jsonBody: try Self.encoder.encode(body)))
    }

    // MARK: WebSocket event stream

    /// Live event stream from `GET /api/ws`. The task is created and resumed when
    /// iteration begins; cancelling the consuming task (or breaking the loop)
    /// tears the socket down via `onTermination`. On any socket failure the
    /// stream finishes with an error so the caller can fall back to polling.
    func events() -> AsyncThrowingStream<EdgeEvent, Error> {
        AsyncThrowingStream { continuation in
            let wsURL: URL
            do {
                wsURL = try makeWebSocketURL()
            } catch {
                continuation.finish(throwing: error)
                return
            }
            var request = URLRequest(url: wsURL)
            if !token.isEmpty {
                request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            }
            let task = URLSession.shared.webSocketTask(with: request)

            func receive() {
                task.receive { result in
                    switch result {
                    case .failure(let error):
                        continuation.finish(throwing: error)
                    case .success(let message):
                        switch message {
                        case .string(let text):
                            Self.emit(text, to: continuation)
                        case .data(let data):
                            if let text = String(data: data, encoding: .utf8) {
                                Self.emit(text, to: continuation)
                            }
                        @unknown default:
                            break
                        }
                        receive()
                    }
                }
            }

            task.resume()
            receive()
            continuation.onTermination = { _ in
                task.cancel(with: .goingAway, reason: nil)
            }
        }
    }

    private static func emit(
        _ text: String,
        to continuation: AsyncThrowingStream<EdgeEvent, Error>.Continuation
    ) {
        // A single frame may carry one or more newline-delimited JSON events.
        for line in text.split(separator: "\n") {
            guard let data = String(line).data(using: .utf8) else { continue }
            if let event = try? decoder.decode(EdgeEvent.self, from: data) {
                continuation.yield(event)
            }
        }
    }

    // MARK: Request plumbing

    private func makeURL(_ path: String) throws -> URL {
        guard let url = URL(string: trimmedHost + path) else {
            throw EdgeError("Invalid AutoPi host URL")
        }
        return url
    }

    private func makeWebSocketURL() throws -> URL {
        guard var comps = URLComponents(string: trimmedHost + "/api/ws") else {
            throw EdgeError("Invalid AutoPi host URL")
        }
        switch comps.scheme {
        case "http": comps.scheme = "ws"
        case "https": comps.scheme = "wss"
        default: break
        }
        if !token.isEmpty {
            comps.queryItems = [URLQueryItem(name: "token", value: token)]
        }
        guard let url = comps.url else {
            throw EdgeError("Invalid AutoPi WebSocket URL")
        }
        return url
    }

    private func makeRequest(_ path: String, method: String, jsonBody: Data? = nil) throws -> URLRequest {
        var request = URLRequest(url: try makeURL(path))
        request.httpMethod = method
        request.timeoutInterval = 15
        if !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let jsonBody {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = jsonBody
        }
        return request
    }

    private func perform<R: Decodable>(_ request: URLRequest) async throws -> R {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw EdgeError("Network error: \(error.localizedDescription)")
        }
        guard let http = response as? HTTPURLResponse else {
            throw EdgeError("Unexpected non-HTTP response")
        }
        guard (200..<300).contains(http.statusCode) else {
            throw EdgeError(Self.message(forStatus: http.statusCode), statusCode: http.statusCode)
        }
        do {
            return try Self.decoder.decode(R.self, from: data)
        } catch {
            throw EdgeError("Could not decode response: \(error.localizedDescription)")
        }
    }

    private static func message(forStatus code: Int) -> String {
        switch code {
        case 401: return "Unauthorized — check the bearer token"
        case 404: return "Not found (404)"
        case 409: return "AutoPi is busy — a job is already running (409)"
        default: return "AutoPi returned HTTP \(code)"
        }
    }
}

File: companion/android/app/src/main/java/com/inomotech/canrosetta/companion/remote/EdgeControlClient.kt
package com.inomotech.canrosetta.companion.remote

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import org.json.JSONObject
import java.io.IOException
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

// MARK: - Wire models (match docs/control-protocol.md)

enum class EdgeMode(val wire: String, val label: String) {
    FAST("fast", "Fast"),
    SLOW("slow", "Slow");

    companion object {
        fun fromWire(value: String?): EdgeMode = entries.firstOrNull { it.wire == value } ?: FAST
    }
}

data class HealthResponse(val ok: Boolean, val swVersion: String?)
data class TimeResponse(val tUtc: Double)
data class EdgeDevice(val id: String?, val swVersion: String?)
data class EdgeStats(val elapsedS: Double?, val frames: Int?, val obdSamples: Int?)
data class DiscoverySummary(val obdPids: Int?, val udsDids: Int?, val plainCanIds: Int?)

data class StatusResponse(
    val state: String,
    val sessionId: String?,
    val outputDir: String?,
    val device: EdgeDevice?,
    val mode: String?,
    val stats: EdgeStats?,
    val discoverySummary: DiscoverySummary?,
    val error: String?,
)

data class SessionResponse(val sessionId: String, val outputDir: String?, val device: EdgeDevice?)
data class CommandResponse(val state: String?, val frames: Int?)

data class EdgeEvent(
    val event: String,
    val state: String? = null,
    val phase: String? = null,
    val supportedPids: Int? = null,
    val summary: DiscoverySummary? = null,
    val frames: Int? = null,
    val obdSamples: Int? = null,
    val elapsedS: Double? = null,
    val message: String? = null,
    val ts: Double? = null,
)

/** A user-presentable error from the control client. */
class EdgeException(message: String, val statusCode: Int? = null) : Exception(message)

/**
 * Client for the AutoPi control protocol (`docs/control-protocol.md`) over
 * OkHttp. Holds only host + token, so it can be recreated cheaply whenever the
 * pairing settings change (the OkHttp clients themselves are shared singletons).
 *
 * HTTP requests carry `Authorization: Bearer <token>`; the WebSocket carries the
 * token both as an `Authorization` header and a `?token=` query param.
 */
class EdgeControlClient(host: String, private val token: String) {

    private val baseUrl: String = host.trim().trimEnd('/')

    companion object {
        private val JSON = "application/json; charset=utf-8".toMediaType()

        private val httpClient: OkHttpClient by lazy {
            OkHttpClient.Builder()
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(15, TimeUnit.SECONDS)
                .callTimeout(20, TimeUnit.SECONDS)
                .build()
        }

        private val wsClient: OkHttpClient by lazy {
            OkHttpClient.Builder()
                .connectTimeout(10, TimeUnit.SECONDS)
                .pingInterval(20, TimeUnit.SECONDS)
                .readTimeout(0, TimeUnit.SECONDS) // keep the socket open indefinitely
                .build()
        }
    }

    // MARK: HTTP endpoints

    suspend fun health(): HealthResponse {
        val o = getJson("/api/health")
        return HealthResponse(o.optBoolean("ok", false), o.optStringOrNull("sw_version"))
    }

    suspend fun time(): TimeResponse = TimeResponse(getJson("/api/time").getDouble("t_utc"))

    suspend fun status(): StatusResponse = parseStatus(getJson("/api/status"))

    suspend fun createSession(
        sessionId: String?,
        edgeUtcOffsetEstS: Double?,
        clockSource: String?,
    ): SessionResponse {
        val body = JSONObject()
        if (sessionId != null) body.put("session_id", sessionId)
        if (edgeUtcOffsetEstS != null) body.put("edge_utc_offset_est_s", edgeUtcOffsetEstS)
        if (clockSource != null) body.put("clock_source", clockSource)
        val o = postJson("/api/session", body)
        return SessionResponse(
            sessionId = o.optString("session_id", sessionId ?: ""),
            outputDir = o.optStringOrNull("output_dir"),
            device = parseDevice(o.optJSONObject("device")),
        )
    }

    suspend fun discover(mode: EdgeMode): CommandResponse =
        parseCommand(postJson("/api/discover", JSONObject().put("mode", mode.wire)))

    suspend fun startLog(): CommandResponse = parseCommand(postJson("/api/log/start", null))

    suspend fun stopLog(): CommandResponse = parseCommand(postJson("/api/log/stop", null))

    suspend fun run(mode: EdgeMode, durationS: Double? = null): CommandResponse {
        val body = JSONObject().put("mode", mode.wire)
        body.put("duration_s", if (durationS != null) durationS else JSONObject.NULL)
        return parseCommand(postJson("/api/run", body))
    }

    // MARK: WebSocket event stream

    /**
     * Live event stream from `GET /api/ws`. Each socket frame may carry one or
     * more newline-delimited JSON events. The flow completes when the socket
     * closes and errors when it fails, so the caller can fall back to polling.
     */
    fun events(): Flow<EdgeEvent> = callbackFlow {
        val listener = object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                for (line in text.split("\n")) {
                    val trimmed = line.trim()
                    if (trimmed.isEmpty()) continue
                    try {
                        trySend(parseEvent(JSONObject(trimmed)))
                    } catch (_: Exception) {
                        // ignore malformed lines
                    }
                }
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                onMessage(webSocket, bytes.utf8())
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                close(t)
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(code, reason)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                close()
            }
        }
        val ws = wsClient.newWebSocket(buildWsRequest(), listener)
        awaitClose { ws.cancel() }
    }

    // MARK: Request plumbing

    private suspend fun getJson(path: String): JSONObject = execute(buildRequest(path, "GET", null))

    private suspend fun postJson(path: String, body: JSONObject?): JSONObject =
        execute(buildRequest(path, "POST", body))

    private fun buildRequest(path: String, method: String, body: JSONObject?): Request {
        val builder = Request.Builder().url(baseUrl + path)
        if (token.isNotEmpty()) builder.header("Authorization", "Bearer $token")
        when (method) {
            "POST" -> builder.post((body?.toString() ?: "{}").toRequestBody(JSON))
            else -> builder.get()
        }
        return builder.build()
    }

    private fun buildWsRequest(): Request {
        val url = buildString {
            append(baseUrl).append("/api/ws")
            if (token.isNotEmpty()) append("?token=").append(URLEncoder.encode(token, "UTF-8"))
        }
        val builder = Request.Builder().url(url)
        if (token.isNotEmpty()) builder.header("Authorization", "Bearer $token")
        return builder.build()
    }

    private suspend fun execute(request: Request): JSONObject = withContext(Dispatchers.IO) {
        val response = try {
            httpClient.newCall(request).execute()
        } catch (e: IOException) {
            throw EdgeException("Network error: ${e.message}")
        }
        response.use { resp ->
            val bodyStr = resp.body?.string().orEmpty()
            if (!resp.isSuccessful) throw EdgeException(messageForStatus(resp.code), resp.code)
            if (bodyStr.isBlank()) JSONObject() else JSONObject(bodyStr)
        }
    }

    private fun messageForStatus(code: Int): String = when (code) {
        401 -> "Unauthorized — check the bearer token"
        404 -> "Not found (404)"
        409 -> "AutoPi is busy — a job is already running (409)"
        else -> "AutoPi returned HTTP $code"
    }

    // MARK: Parsing

    private fun parseDevice(o: JSONObject?): EdgeDevice? =
        o?.let { EdgeDevice(it.optStringOrNull("id"), it.optStringOrNull("sw_version")) }

    private fun parseSummary(o: JSONObject?): DiscoverySummary? =
        o?.let {
            DiscoverySummary(
                it.optIntOrNull("obd_pids"),
                it.optIntOrNull("uds_dids"),
                it.optIntOrNull("plain_can_ids"),
            )
        }

    private fun parseStatus(o: JSONObject): StatusResponse = StatusResponse(
        state = o.optString("state", "unknown"),
        sessionId = o.optStringOrNull("session_id"),
        outputDir = o.optStringOrNull("output_dir"),
        device = parseDevice(o.optJSONObject("device")),
        mode = o.optStringOrNull("mode"),
        stats = o.optJSONObject("stats")?.let {
            EdgeStats(it.optDoubleOrNull("elapsed_s"), it.optIntOrNull("frames"), it.optIntOrNull("obd_samples"))
        },
        discoverySummary = parseSummary(o.optJSONObject("discovery_summary")),
        error = o.optStringOrNull("error"),
    )

    private fun parseCommand(o: JSONObject): CommandResponse =
        CommandResponse(o.optStringOrNull("state"), o.optIntOrNull("frames"))

    private fun parseEvent(o: JSONObject): EdgeEvent = EdgeEvent(
        event = o.optString("event", ""),
        state = o.optStringOrNull("state"),
        phase = o.optStringOrNull("phase"),
        supportedPids = o.optIntOrNull("supported_pids"),
        summary = parseSummary(o.optJSONObject("summary")),
        frames = o.optIntOrNull("frames"),
        obdSamples = o.optIntOrNull("obd_samples"),
        elapsedS = o.optDoubleOrNull("elapsed_s"),
        message = o.optStringOrNull("message"),
        ts = o.optDoubleOrNull("ts"),
    )
}

// org.json helpers that return null (not sentinel defaults) for absent/null keys.
private fun JSONObject.optStringOrNull(key: String): String? =
    if (has(key) && !isNull(key)) optString(key) else null

private fun JSONObject.optIntOrNull(key: String): Int? =
    if (has(key) && !isNull(key)) optInt(key) else null

private fun JSONObject.optDoubleOrNull(key: String): Double? =
    if (has(key) && !isNull(key)) optDouble(key) else null
